File: composeApp/src/iosMain/kotlin/org/example/kmpsamples/presentation/permissions/systemPermissionHandlers/RecordAudioHandler.kt
package org.example.kmpsamples.presentation.permissions.systemPermissionHandlers

import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import org.example.kmpsamples.permissions.systemPermissionHandlers.SystemPermissionHandler
import org.example.kmpsamples.presentation.permissions.dtos.Permission
import org.example.kmpsamples.presentation.permissions.dtos.PermissionState
import org.example.kmpsamples.presentation.permissions.dtos.PermissionStateType
import org.example.kmpsamples.presentation.permissions.dtos.PermissionStateType.SinglePermissionState
import org.example.kmpsamples.presentation.permissions.dtos.PlatformPermissionType.PlatformSinglePermission
import platform.AVFAudio.AVAudioSession
import platform.AVFAudio.AVAudioSessionRecordPermissionDenied
import platform.AVFAudio.AVAudioSessionRecordPermissionGranted
import platform.AVFAudio.AVAudioSessionRecordPermissionUndetermined

class RecordAudioHandler() : SystemPermissionHandler {
    private val mainScope = MainScope()

    override fun canHandle(simplePlatformPermission: PlatformSinglePermission): Boolean {
        fun isSupported(permission: Permission): Boolean = when (permission) {
            is Permission.RecordAudioPermission -> true
            else -> false
        }
        return isSupported(simplePlatformPermission.permission)
    }

    private fun createSimplePermissionState(
        isGranted: Boolean,
        shouldShowRationale: Boolean
    ): SinglePermissionState {
        val permissionState =
            PermissionState(Permission.RecordAudioPermission(), isGranted, shouldShowRationale)
        return SinglePermissionState(permissionState)
    }

    override fun resolvePermissionState(): PermissionStateType {
        val status = AVAudioSession.sharedInstance().recordPermission
        return when (status) {
            AVAudioSessionRecordPermissionGranted -> createSimplePermissionState(true, false)
            AVAudioSessionRecordPermissionDenied -> createSimplePermissionState(false, false)
            AVAudioSessionRecordPermissionUndetermined -> createSimplePermissionState(false, true)
            else -> createSimplePermissionState(false, true)
        }
    }

    override fun requestPermission(onRequest: (permissionState: PermissionStateType) -> Unit) {
        AVAudioSession.sharedInstance().requestRecordPermission { isGranted ->
            mainScope.launch {
                onRequest(resolvePermissionState())
            }
        }
    }
}

File: iosApp/iosApp/Permissions/SystemPermissionHandlers/RecordAudioHandler.swift
import AVFAudio
import ComposeApp

class RecordAudioHandler: SystemPermissionHandler {
    var id: String = "RecordAudioHandler"
    
    private let mainScope = DispatchQueue.main
    
    func canHandle(simplePlatformPermission: PlatformPermissionTypePlatformSinglePermission) -> Bool {
        return simplePlatformPermission.permission is Permission.RecordAudioPermission
    }
    
    private func createSimplePermissionState(
        isGranted: Bool,
        shouldShowRationale: Bool
    ) -> PermissionStateTypeSinglePermissionState {
        let permissionState = PermissionState(permission: Permission.RecordAudioPermission(), isGranted: isGranted, shouldShowRationale: shouldShowRationale)
        return PermissionStateTypeSinglePermissionState(permissionState: permissionState)
    }
    
    func resolvePermissionState() -> PermissionStateType {
        let status = AVAudioSession.sharedInstance().recordPermission
        switch status {
        case .granted:
            return createSimplePermissionState(isGranted: true, shouldShowRationale: false)
        case .denied:
            return createSimplePermissionState(isGranted: false, shouldShowRationale: false)
        case .undetermined:
            return createSimplePermissionState(isGranted: false, shouldShowRationale: true)
        @unknown default:
            return createSimplePermissionState(isGranted: false, shouldShowRationale: true)
        }
    }
    
    func requestPermission(onRequest: @escaping (PermissionStateType) -> Void) {
        AVAudioSession.sharedInstance().requestRecordPermission { isGranted in
            self.mainScope.async {
                onRequest(self.resolvePermissionState())
            }
        }
    }
}
